keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1248 @@
+
+An out-of-tree CSI plugin can provide flexibility and modularity, but there are trade-offs to consider:
+
+ - Complexity of managing an external CSI plugin. This includes handling the installation, configuration, and updates of the CSI driver, which adds
+   an additional operational burden. For a generic, vendor-agnostic, and widely-adopted solution this would not make sense.
+ - Supporting the image pull secrets as well as credentials provider will be tricky and needs to be reimplemented with the separate API calls.
+ - External CSI plugins implement their own lifecycle management and garbage collection mechanisms,
+   yet these already exist in-tree for OCI images.
+    - The kubelet has max parallel image pull constant to maintain the reasonable
+      load on a disk and network. This will not be respected by CSI driver and
+      the only point of integration may be if we move this constant down to
+      runtime.
+    - The kubelet has GC logic that is not cleaning up images immediately in
+      case they will be reused. Also GC logic has it's own thresholds and
+      behavior on eviction. It will be nice to have those integrated.
+    - The kubelet exposes metrics on image pulls and we have KEP in place to
+      improve it even further. Having CSI exposing those metrics will require
+      customer to integrate with one more source of data.
+ - Performance: There is additional overhead with an out-of-tree CSI plugin, especially in scenarios requiring frequent image pulls
+   or large volumes of data.
+
+### Advantages of In-Tree OCI VolumeSource
+
+1. **Leverage Existing Mechanisms:**
+   - **No New Data Types or Objects:** OCI images are already a core part of the Kubernetes ecosystem. Extending support for OCI artifacts, many of
+     the same mechanisms will be reused. This ensures consistency and reduces complexity, as both adhere to the same OCI image format.
+   - **Existing Lifecycle Management and Garbage Collection:** Kubernetes has efficient lifecycle management and garbage collection mechanisms for
+	 volumes and container images. The in-tree OCI VolumeSource will utilize these existing mechanisms.
+
+2. **Integration with Kubernetes:**
+   - **Optimal Performance:** Deep integration with the scheduler and kubelet ensures optimal performance and
+     resource management. This integration allows the OCI VolumeSource to benefit from all existing optimizations and features.
+   - **Unified Interface:** Users interact with a consistent and unified interface for managing volumes, reducing the learning curve and
+     potential for configuration errors.
+
+3. **Simplified Maintenance and Updates:**
+   - **Core Project Maintenance:** In-tree features are maintained and updated as part of the core project. It makes sense
+     for widely-used and vendor agnostic features to utilize the core testing infrastructure, release cycles, and security updates.
+
```

Comment:
The one obvious alternative that was not even listed here is using shared volumes inside a single pod. Was this considered? What were the reasons for a new volume source?

Reply:
Yup, this is what prompted the proposal - I was annoyed because I wanted to mount  an LLM as a separate image rather than combine an LLM + model-server into 1 image (for modularity/easily plugging in different models). In order to do this, I had to package the LLM in an image with a shell in its base layer. Then, an init container to copy the LLM (using the linux `install` command) from the model image to a volume shared with the model-server container. In order to remove the need for a shell & init container, I propose a new OCI VolumeSource. That alternative _should_ be explicitly stated here, I'll add it!

Reply:
If you mount and ORAS artifact you can do away with your LLM data also needing a filesystem.

Reply:
I assume this conversation is resolved with https://github.com/kubernetes/enhancements/pull/4642/commits/e5013dc14aba730b4efad49453fb821d0ecbd551, feel free to reopen if I've missed anything.

Reply:
This still holds, see my other comment.